### PR TITLE
Build with BerkeleyDB >= 4.6

### PR DESCRIPTION
**Pull request by sizeofvoid**
Merge OpenBSD patches[1] to unbreak build on OpenBSD and systems with BerkeleyDB >= 4.6. Original patches from me, bitcoin openbsd ports maintainer.

[1]: https://cvsweb.openbsd.org/ports/net/bitcoin/patches/

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: very nice to get this merged upstream!

Great first-time contribution!

**Comment by practicalswift** [comment]
@sizeofvoid `--with-incompatible-bdb` will still be required to build with a bdb version other than 4.8, right?

File: src/wallet/db.cpp [hunk 1]
```diff
@@ -213,7 +215,9 @@ void BerkeleyEnvironment::MakeMock()
     dbenv->set_lk_max_locks(10000);
     dbenv->set_lk_max_objects(10000);
     dbenv->set_flags(DB_AUTO_COMMIT, 1);
+#if DB_VERSION_MINOR > 7
```
[the remark this answers is not in the document]
Comment (sizeofvoid) [reply src/wallet/db.cpp]: Yes, also a good idea!

**Comment by achow101** [comment]
Tend to NACK. We only officially support BDB 4.8 and make no guarantee to support any other version. I don't see why it is necessary to support an even older version of BDB.

**Comment by Cobra-Bitcoin** [comment]
NACK as is. It shouldn't be possible to build with anything other than 4.8, except through `--with-incompatible-bdb`.

File: build-aux/m4/bitcoin_find_bdb48.m4 [hunk 2]
```diff
@@ -35,7 +35,7 @@ AC_DEFUN([BITCOIN_FIND_BDB48],[
       AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
         #include <${searchpath}db_cxx.h>
       ]],[[
-        #if !(DB_VERSION_MAJOR == 4 && DB_VERSION_MINOR == 8)
+        #if !(DB_VERSION_MAJOR == 4 && DB_VERSION_MINOR == 6)
           #error "failed to find bdb 4.8"
```
[the remark this answers is not in the document]
Comment (promag) [reply build-aux/m4/bitcoin_find_bdb48.m4]: And the file should be renamed..

File: src/wallet/db.cpp [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2009-2010 Satoshi Nakamoto
+
```
Comment (promag) [inline src/wallet/db.cpp]: Why this change?

File: src/wallet/db.cpp [hunk 4]
```diff
@@ -156,7 +156,9 @@ bool BerkeleyEnvironment::Open(bool retry)
     dbenv->set_errfile(fsbridge::fopen(pathErrorFile, "a")); /// debug
     dbenv->set_flags(DB_AUTO_COMMIT, 1);
     dbenv->set_flags(DB_TXN_WRITE_NOSYNC, 1);
+#if DB_VERSION_MINOR > 7
```
[the remark this answers is not in the document]
Comment (promag) [reply src/wallet/db.cpp]: > It shouldn't be possible to build with anything other than 4.8, except through `--with-incompatible-bdb`

Is this required to build with < 4.8?

**Comment by sizeofvoid** [comment]
> @sizeofvoid `--with-incompatible-bdb` will still be required to build with a bdb version other than 4.8, right?

On OpenBSD 6.3 and upcoming 6.4 with db-4.6.21 it fails also with `--with-incompatible-bdb`:
```
checking for Berkeley DB C++ headers... no
configure: error: libdb_cxx headers missing, Bitcoin Core requires this library for wallet functionality (--disable-wallet to disable wallet functionality)                                   
*** Error 1 in . (/usr/ports/infrastructure/mk/bsd.port.mk:2693 'do-configure')
*** Error 1 in . (/usr/ports/infrastructure/mk/bsd.port.mk:2713 '/usr/ports/pobj/bitcoin-0.17.0/.configure_done')                                                                             
*** Error 1 in /usr/ports/net/bitcoin (/usr/ports/infrastructure/mk/bsd.port.mk:2419 'configure') 
```

File: src/wallet/db.cpp (hunk 3 above)
Comment (sizeofvoid) [reply src/wallet/db.cpp]: Of course it shouldn't be.

**Comment by fanquake** [comment]
Agree with the other comments, we shouldn't be making code-changes to allow for building with < 4.8. We also already provide a [db4 install script](https://github.com/bitcoin/bitcoin/blob/master/contrib/install_db4.sh), which is the way we recommended installing BerkeleyDB in the [OpenBSD docs](https://github.com/bitcoin/bitcoin/blob/master/doc/build-openbsd.md#building-berkeleydb).

**Comment by practicalswift** [comment]
@sizeofvoid It seems like the consensus is NACK. Does keeping this patch in the ports system cause any trouble for the OpenBSD project? :-)
